File: PascalsTriangle/PascalsTrianglev1.java
package PascalsTriangle;

import java.util.*;

public class PascalsTrianglev1 {
    public static void main(String[] args) {
        Scanner kbd = new Scanner(System.in);
        boolean isValid = false;

        while (!isValid) {
            try {
                System.out.print("Enter the number of rows desired for Pascal's Triangle: ");
                int rows = kbd.nextInt();
                
                if (rows <= 0) {
                    System.out.println("Please enter a positive integer.");
                    continue; // Ask for a valid value again
                }

                List<List<Integer>> pascalTriangle = generate(rows);
                for (List<Integer> row : pascalTriangle) {
                    System.out.println(row);
                }

                isValid = true; // A valid value is entered, exit the loop
            } catch (InputMismatchException e) {
                System.out.println("Invalid input. Please enter a positive integer.");
                kbd.next(); // Consume the invalid token
            }
        }

        kbd.close(); // Close the scanner at the end
    }

    public static List<List<Integer>> generate(int numRows) {
        List<List<Integer>> result = new ArrayList<>(); // Main list
    
        if (numRows <= 0) {return result;} // If there are no rows, return an empty list
    
        // The first row of Pascal's Triangle is always [1]
        result.add(new ArrayList<>(List.of(1)));
    
        // Build rows from the second to numRows
        for (int i = 1; i < numRows; i++) {
            List<Integer> prevRow = result.get(i - 1); // Previous row
            List<Integer> currentRow = new ArrayList<>();
    
            // First element of each row is 1
            currentRow.add(1);
    
            // Build the intermediate elements
            for (int j = 1; j < prevRow.size(); j++) {
                currentRow.add(prevRow.get(j - 1) + prevRow.get(j));
            }
    
            // Last element of each row is 1
            currentRow.add(1);
    
            // Add the current row to the result
            result.add(currentRow);
        }
    
        return result;
    }
}

File: PascalsTriangle/PascalsTrianglev1.py
def generate(num_rows):
    result = []  # Lista principal

    if num_rows <= 0:
        return result  # Si no hay filas, devolver lista vacía

    # Primera fila siempre [1]
    result.append([1])

    # Construir las filas desde la segunda hasta num_rows
    for i in range(1, num_rows):
        prev_row = result[i - 1]  # Fila anterior
        current_row = [1]  # Primer elemento siempre 1

        # Construir los elementos intermedios usando la fila anterior
        for j in range(1, len(prev_row)):
            current_row.append(prev_row[j - 1] + prev_row[j])

        current_row.append(1)  # Último elemento siempre 1

        result.append(current_row)  # Agregar la nueva fila al resultado

    return result


def main():
    is_valid = False
    while not is_valid:
        try:
            rows = int(input("Enter the number of rows desired for Pascal's Triangle: "))
            if rows <= 0:
                print("Please enter a positive integer.")
                continue  # Pedir de nuevo un valor válido

            pascal_triangle = generate(rows)
            for row in pascal_triangle:
                print(row)

            is_valid = True  # Valor válido → salir del bucle
        except ValueError:
            print("Invalid input. Please enter a positive integer.")


if __name__ == "__main__":
    main()
